### PR TITLE
Email Notification System Phase 1

### PR DESCRIPTION
**Pull request by SianaDicheva**
# Description 

The following pages were created in order to ensure a nice creation flow of email notifications and edit flow: 

Notification List and Distribution List Page 

This page allows users to create, edit, delete, clone and send emails directly to the distribution list that was assigned to each notification. The code is located in the email folder in components.

You can find code for this in the: 
Ems-template Folder 
Email Root Folder 

Step 1- Create Notification Page 

This page allows users to set the name of the notification they have created and checks if a notification with that name already exists (if in edit mode, this isn’t the case. 

You can find code for this in the: Create-notification Folder 

Step 2 – Select Dataset Page 

This page is used to allow the user to create blocks and filter and select fields from data to form a table which can be selected in the layout step and then be seen in the final sent email. 

You can find code for this in the: Create-dataset, Dataset-Filter, Filter, Email/graphql 

Step 3 - Select Distribution List Page 

This page allows users to select an existing distribution list or create a new and from here the user can filter distribution lists from blocks and select the to recipients, cc and bcc recipients. 

You can find the code for this in the: Select Distribution List Folder, Email-template Folder, Filter, Email/GraphQL 

Step 4 – Schedule Alert (Coming in next release) 

This step will be in the next release. Will be used to be able to schedule alerts. 

Step 5 - Layout Page 

This page allows users to customise the final email look. Add subject text with tokens(e.g. {{today.now}}), header, footer, body with relevant Blocks, attach images and a banner plus styling of the final email.

You can find the code for this in the: Layout Folder, Email Service 

Step 6 - Preview Page 

This page allows users to view the preview before its sent. Two options - Save or Send, where Send saves it and sends the email.

You can find the code for this in the: Preview Folder, Email Service 

Shared components:  

email service - This is used to share data between different pages and retrieve data from the database. 

Email component - Used to send the email and save data to the database. 

 
## Useful links

- Please insert link to ticket - https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/16252
- Please insert link to back-end branch - https://github.com/ReliefApplications/ems-backend/pull/962

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

### CODE REVIEW

**Comment by MwanPygmay** [comment]
Hi  @SianaDicheva, here are some comments that I have for this PR, feel free to tag me if anything is not clear!

TLDR: 
- Add some comments, you can see all missing comment by running npm run lint.
- Try to use more the components that were already created and that fit the current application styling, like ui-tabs, shared-tab-fields...

Some more detailed things that caught my eye:
- Getting a spinner even though the page appears to be loaded when creating a new notification
- Missing documentation for most variables: all missing doc can be seen by runnig 'npm run lint'
- Rather than using kendo-tabstrip, I think you could rather use the ui-tabs component that we use in other places of the code
- In the same logic,  you can use the shared-tab-filter and the shared-tab-fields for the 'Filter' and 'Fields' part of the dataset-filter (I think it should make the styling more consistent across the app).
- If you want an example of how we use the shared-tab-filter, the shared-tab-fields and the ui-tabs, go in a dashboard, create a grid, select a resource, create a new layout => there we use the query-builder which I think looks about the same of what you started to do for the dataset-filter.
- For the alert with the class 'validation-note', I think it would be better to use the ui-alert component
- For the part with the 'To', 'cc' etc, we already have an e-mail template that looks a bit like it, you can take a look at it in the shared-email-preview-modal
- the 'Alert me' button is not aligned
- For the 'Schedule alert', I think you may rely on using the ui-cron-editor that does pretty much what you need.
- In the 'Layout' part, please try to inspire from the email template we already use for better consistency. Also, there is no warning or anything that warns the user about what is missing to be able to save and proceed.
- Repair case: in some places, the case feels weird, for instance 'Select Block Field' should be 'Select block field'.
- Maybe this is normal but I cannot Save and proceed the layout part and cannot go to the preview, don't know what is missing.


cc @AntoineRelief

**Comment by SianaDicheva** [comment]
> Hi @SianaDicheva, here are some comments that I have for this PR, feel free to tag me if anything is not clear!
> 
> TLDR:
> 
> * Add some comments, you can see all missing comment by running npm run lint.
> * Try to use more the components that were already created and that fit the current application styling, like ui-tabs, shared-tab-fields...
> 
> Some more detailed things that caught my eye:
> 
> * Getting a spinner even though the page appears to be loaded when creating a new notification
> * Missing documentation for most variables: all missing doc can be seen by runnig 'npm run lint'
> * Rather than using kendo-tabstrip, I think you could rather use the ui-tabs component that we use in other places of the code
> * In the same logic,  you can use the shared-tab-filter and the shared-tab-fields for the 'Filter' and 'Fields' part of the dataset-filter (I think it should make the styling more consistent across the app).
> * If you want an example of how we use the shared-tab-filter, the shared-tab-fields and the ui-tabs, go in a dashboard, create a grid, select a resource, create a new layout => there we use the query-builder which I think looks about the same of what you started to do for the dataset-filter.
> * For the alert with the class 'validation-note', I think it would be better to use the ui-alert component
> * For the part with the 'To', 'cc' etc, we already have an e-mail template that looks a bit like it, you can take a look at it in the shared-email-preview-modal
> * the 'Alert me' button is not aligned
> * For the 'Schedule alert', I think you may rely on using the ui-cron-editor that does pretty much what you need.
> * In the 'Layout' part, please try to inspire from the email template we already use for better consistency. Also, there is no warning or anything that warns the user about what is missing to be able to save and proceed.
> * Repair case: in some places, the case feels weird, for instance 'Select Block Field' should be 'Select block field'.
> * Maybe this is normal but I cannot Save and proceed the layout part and cannot go to the preview, don't know what is missing.
> 
> cc @AntoineRelief

Hi @MwanPygmay , 

Thank you for the comment!

We will be adding a few more changes today and we will definitely look into your comments and suggestions.

**Comment by HasnatL-Adapptlimited** [comment]
Hey @MwanPygmay, CC: @matheus-relief, @AntoineRelief
Thank you for the feedback, we have reviewed your comments, some points:
 
#### In the same logic, you can use the shared-tab-filter and the shared-tab-fields for the 'Filter' and 'Fields' part of the dataset-filter (I think it should make the styling more consistent across the app) 
#### &&
#### Rather than using kendo-tabstrip, I think you could use the ui-tabs component that we use in other places of the code
 
- We chose to pluck the code from shared-tab-filter + shared-tab-field but didn't place the whole templates in as:
- We added a new operator called "within the last" (takes the date today and filters by dates in the last 'x hours/days/month/ years' so it will most likely break everywhere that shared-filter is used)
 - We wanted to keep the email notification code isolated from the rest of the app builder code so that in future it can be taken out and implemented in other applications instead of using shared components (Similar to what we have done with the ui-buttons, ui-select, by inheriting the styles but not fully using the templates)
 
#### For the part with the 'To', 'cc' etc, we already have an e-mail template that looks a bit like it, you can take a look at it in the shared-email-preview-modal + In the 'Layout' part, please try to inspire from the email template we already use for better consistency.

##### Distribution lists have become more complex:
- We have extra functionality to include to, cc and bcc
 - Provided more options to choose from datasets, reference data and apply filters on top of it which cannot be done with the existing application, which is why we went with our design.

 
## Things that have been fixed:
 
- Getting a spinner even though the page appears to be loaded when creating a new notification - Fixed
 - Maybe this is normal but I cannot Save and proceed the layout part and cannot go to the preview, don't know what is missing. - Fixed
- For the alert with the class 'validation-note', I think it would be better to use the ui-alert component - Fixed
- Missing documentation for most variables: all missing doc can be seen by running 'npm run lint' - 200 warnings - Fixed
- Repair case: in some places, the case feels weird, for instance, 'Select Block Field' should be 'Select block field'. - Believe Fixed -- but need to verify how it should be as in some places of the app builder cases are not in a uniform manner.

## In progress:
- the 'Alert me' button is not aligned - Hidden for now as functionality is for phase 2
 - For the 'Schedule alert', I think you may rely on using the ui-cron-editor which does pretty much what you need. -- **Will Be Released as Part of Phase 2**

**Comment by AntoineRelief** [comment]
hey @HasnatL-Adapptlimited  

Sorry took me a while to come back to this ticket with the SUI release

about this comment:

> We wanted to keep the email notification code isolated from the rest of the app builder code so that in future it can be taken out and implemented in other applications instead of using shared components (Similar to what we have done with the ui-buttons, ui-select, by inheriting the styles but not fully using the templates)

I'm really not sure, because as it is, it looks quite different from the rest of the UI
We've done a lot of work past year to create new components in order to have a modern UI, and make sure the platform would look like an unique product, not a combination of multiplie libraries

In addition, using shared components also help for development, as, if we find issues in them, resolving them affect the whole platform
the shared library can be taken as it is, a library, and be used by other systems if needed, as it can be exported as a npm library if needed

is that a decision you discussed with WHO?

cc @MwanPygmay

**Comment by MwanPygmay** [comment]
Hi @HasnatL-Adapptlimited @SianaDicheva , 
I see some problems with the emails sending: the custom colours are not taken into account, the subject using the {{}} syntax works before sending, but seems to not be applied in the back.
Screenshot_20240227_105211
The custom colours seem to work once but once regoing to the layout page, the previously saved colours are reset to default.
The layout we receive after sending in the email does not fit the original layout:
original (misses some alignments and some blank space does feil weird)
image
layout in the mail (seems very horizontal to me, svg image is not sent, colours are not sent):
image

**Review by MwanPygmay — CHANGES_REQUESTED** [review]
Hi @SianaDicheva @HasnatL-Adapptlimited, sorry Matthis told me you were awaiting for my review. I did not know whether your commits were ready for review or not. To avoid this confusion I will from now put the ticket on Draft mode after reviewing, and you can then put it back in Ready for review once you want me to comment again. I think that would simplify things from my side, do you agree?
As for my comments on this PR:

- The emails received are indeed greatly improved but styling is still not sent (it works in preview mode but upon receiving the email styling does not exist).
- After creating a style, when going to the Layout section the style is discarded:
Peek 2024-03-19 14-57
- The image from the header is not inside of the header: 
image
- As for previous comments, I think the part related to coherence inside of the application is not done yet (please refer to comment from Antoine above). For instance, this component does not look at all like the one from the query-builder (use of drag and drop instead of click):
image
 => I think that for two components sharing a similar display, the use of the same UX/UI is necessary for end users. That's why I think you should be using more existing components from the shared library.

Thanks a lot for your work,
Matthieu

**Comment by MwanPygmay** [comment]
@SianaDicheva @HasnatL-Adapptlimited  sorry I did not mention during the meeting but I think you could also merge the beta branch to get latest changes that we did (I believe it was not done in some time, and some stuff has changed from our side, maybe this could help you with the ui-tabs for example)

**Comment by AntoineRelief** [comment]
@SianaDicheva @HasnatL-Adapptlimited 

cc @MwanPygmay 

I only noticed that when deploying the code to UAT, but it appears that you removed some elements from the UI:

Capture d’écran 2024-03-25 à 16 09 53

and grouped them in the same view:

Capture d’écran 2024-03-25 à 16 10 31

Now the issue I see is that all the elements that would exist before in these views ( example taken from SIT ):
Capture d’écran 2024-03-25 à 16 11 31

you can't see them in UAT:
Capture d’écran 2024-03-25 à 16 12 28


and the two environments share the same data
these elements are supposed to be usable there ( in grid widget configuration ):

Capture d’écran 2024-03-25 à 16 13 28

but with the new notifications, the distribution lists & notifications created in the new view are not selectable there

Was it made on purpose and are we missing some code to make it work?

**Comment by AntoineRelief** [comment]
@SianaDicheva @HasnatL-Adapptlimited 

cc @Matthis-M-ReliefApps @MwanPygmay 

I discussed a bit with WHO about the UI
Having an UI that looks like it's part of the rest of the system is not considered as critical to them, as long as it's only in the back-office

About this point, I'm not entirely sure if it's only for back-office or if it's for both offices. Please confirm this point
Currently, the page appears in the front-office as well, but is not usable, so I'm not sure if it should or not

In any case, I'm putting there some mockups I did, about how the UI could be improved
If you have resources to work on that, that would be great, otherwise, we will see what we can do during the final review

Of course, if you have any opinion about the mockups, please don't hesitate to discuss them

The only point I'd really like to see fixed from these mockups is the usage of ng-select, which I would like to avoid
In case you need any help about that, please let us know, and we can see how to change that

Main page

Notification page


@Matthis-M-ReliefApps @MwanPygmay 
the link to the mockups in confluence:
https://oortcloud.atlassian.net/wiki/spaces/OORT/pages/138084354/MOCK+-+Email+Notifications

**Comment by AntoineRelief** [comment]
@HasnatL-Adapptlimited 

closing this one as it's outdated
